homework/index.js

```diff
@@ -30,18 +30,104 @@
+    const details = [
+      { title: 'Repository', valueOfName: selectedRepository.name },
+      { title: 'Description', value: selectedRepository.description },
+      { title: 'Forks', value: selectedRepository.forks },
+      { title: 'Updated', value: new Date(selectedRepository.updated_at).toLocaleString() },
+    ];
+    details.forEach(detail => {
+      const tr = createAndAppend('tr', tbody);
+      createAndAppend('td', tr, { text: detail.title, class: `label ` });
+      const linkToRepository = createAndAppend('a', tr, {
+        href: selectedRepository.html_url,
+        target: '_blank',
+      });
+
+      createAndAppend('td', linkToRepository, {
+        text: detail.valueOfName,
+        class: 'repository-data',
+      });
+      createAndAppend('td', tr, { text: detail.value, class: 'repository-data' });
+    });
+  }
+
+  function createContributors(selectedRepository, mainContainerForRepository) {
+    const root = document.getElementById('root');
+    const contributorsContainer = createAndAppend('div', mainContainerForRepository, {
+      class: 'right-div',
+    });
+    createAndAppend('h3', contributorsContainer, {
+      text: 'Contributors',
+      class: 'contributors-header',
+    });
+    const ul = createAndAppend('ul', contributorsContainer, { class: 'contributor-list' });
+    fetchJSON(selectedRepository.contributors_url, (err, contributors) => {
+      if (err) {
+        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
+        return;
+      }
+      contributors.forEach(contributor => {
+        const li = createAndAppend('li', ul, { class: 'contributor-item' });
+        const alink = createAndAppend('a', li, { href: contributor.html_url, target: '_blank' });
+        const dataDiv = createAndAppend('div', alink, { class: 'contributor' });
+
+        createAndAppend('img', dataDiv, {
+          class: 'contributor-item',
+          src: contributor.avatar_url,
+          height: 52,
+        });
+        const contributorData = createAndAppend('div', dataDiv, { class: 'contributor-data' });
+        createAndAppend('div', contributorData, {
+          class: 'contributor-login',
+          text: contributor.login,
+        });
+        createAndAppend('div', contributorData, {
+          class: 'contributor-badge',
+          text: contributor.contributions,
+        });
+      });
+    });
+  }
+
   function main(url) {
     fetchJSON(url, (err, data) => {
       const root = document.getElementById('root');
       if (err) {
         createAndAppend('div', root, { text: err.message, class: 'alert-error' });
-      } else {
-        createAndAppend('pre', root, { text: JSON.stringify(data, null, 2) });
+        return;
       }
+      const header = createAndAppend('header', root, { class: 'header' });
+      const mainContainerForRepository = createAndAppend('div', root, { class: 'container' });
+
+      function createHeader(repositories) {
```

Comment:
You have nested this function inside the `main` function. You have done this because you are accessing variables inside the scope of the containing `main` function. The problem with that is that it makes it not immediately clear when reading the code what these dependencies on the outer scope variables are. I would recommend that you move this function outside of the `main()` function to make these dependencies explicit. If you that in VSCode you will red underlines of the variables that are now undefined:

![image](https://user-images.githubusercontent.com/2788771/60960094-a896d080-a309-11e9-93e2-a34297670752.png)


The solution is easy: just pass the missing variables as parameters to the function:

![image](https://user-images.githubusercontent.com/2788771/60960160-c106eb00-a309-11e9-9d5b-87deaab7252f.png)

Now, in your `main()` function call `createHeader()` with the extra parameters needed:

```js
function main(url) {
  fetchJSON(url, (err, data) => {
    const root = document.getElementById('root');
    if (err) {
      createAndAppend('div', root, { text: err.message, class: 'alert-error' });
      return;
    }
    const header = createAndAppend('header', root, { class: 'header' });
    const container = createAndAppend('div', root, { class: 'container' });

    createHeader(data, header, container);
    createDescription(data[0], container);
    createContributors(data[0], container);
  });
}
```

Your `main()` function is now shorter and uncluttered by the nested function. And your `createHeader()` function gets everything it needs explicitly through its parameters list, taking all the guessing away.

Reply:
thanks , very clear explanation .